src/gridfs/upload.ts

```diff
@@ -159,76 +189,15 @@ export class GridFSBucketWriteStream extends Writable implements NodeJS.Writable
     this.state.aborted = true;
     await this.chunks.deleteMany({ files_id: this.id });
   }
-
-  /**
-   * Tells the stream that no more data will be coming in. The stream will
-   * persist the remaining data to MongoDB, write the files document, and
-   * then emit a 'finish' event.
-   *
-   * @param chunk - Buffer to write
-   * @param encoding - Optional encoding for the buffer
-   * @param callback - Function to call when all files and chunks have been persisted to MongoDB
-   */
-  override end(): this;
-  override end(chunk: Buffer): this;
-  override end(callback: Callback<GridFSFile | void>): this;
-  override end(chunk: Buffer, callback: Callback<GridFSFile | void>): this;
-  override end(chunk: Buffer, encoding: BufferEncoding): this;
-  override end(
-    chunk: Buffer,
-    encoding: BufferEncoding | undefined,
-    callback: Callback<GridFSFile | void>
-  ): this;
-  override end(
-    chunkOrCallback?: Buffer | Callback<GridFSFile | void>,
-    encodingOrCallback?: BufferEncoding | Callback<GridFSFile | void>,
-    callback?: Callback<GridFSFile | void>
-  ): this {
-    const chunk = typeof chunkOrCallback === 'function' ? undefined : chunkOrCallback;
-    const encoding = typeof encodingOrCallback === 'function' ? undefined : encodingOrCallback;
-    callback =
-      typeof chunkOrCallback === 'function'
-        ? chunkOrCallback
-        : typeof encodingOrCallback === 'function'
-        ? encodingOrCallback
-        : callback;
-
-    if (this.state.streamEnd || checkAborted(this, callback)) return this;
-
-    this.state.streamEnd = true;
-
-    if (callback) {
-      this.once(GridFSBucketWriteStream.FINISH, (result: GridFSFile) => {
-        if (callback) callback(undefined, result);
-      });
-    }
-
-    if (!chunk) {
-      waitForIndexes(this, () => !!writeRemnant(this));
-      return this;
-    }
-
-    this.write(chunk, encoding, () => {
-      writeRemnant(this);
-    });
-
-    return this;
-  }
 }
 
-function __handleError(
-  stream: GridFSBucketWriteStream,
-  error: AnyError,
-  callback?: Callback
-): void {
+function handleError(stream: GridFSBucketWriteStream, error: Error, callback: Callback): void {
   if (stream.state.errored) {
+    process.nextTick(callback);
```

Comment:
That is more of a style/correctness improvement. Callbacks shouldn't be invoked synchronously within the function that was given the callback because it means it won't return until the callback has finished running.

```js
stream._write(chunk, encoding, (err) => {
  console.log('finished write')
  for (;;) {}
})
console.log('started write')
```

Take the code above, because `_write` is a callback-accepting function it should always return once it is done setting up the write, and the callback is invoked when the write is done. If `_write` doesn't move the callback's execution to a separate tick it is including the callback's execution in its own execution. In this example, we'd never reach "started write" even though we have started and potentially finished the write.

Reply:
Thanks for the explanation.